src/main/java/bio/terra/common/ErrorCollector.java

```diff
@@ -20,14 +20,16 @@ public boolean anyErrorsCollected() {
   }
 
   public void record(String lineErrorMsgFormat, Object... lineErrorMsgArgs) {
-    if (errors.size() < maxErrorsReported) {
-      errors.add(String.format(lineErrorMsgFormat, lineErrorMsgArgs));
-    } else if (errors.size() == maxErrorsReported) {
-      errors.add(
-          "Error details truncated. [MaxBadLoadFileLineErrorsReported = "
-              + maxErrorsReported
-              + "]");
-      throw getFormattedException();
+    synchronized (errors) {
```

Comment:
This seems like a heavyweight solution to a concurrency problem. Adding a `boolean` field like `truncationHasBeenReported` and using that instead of `== maxErrorsReported` would be a lighter weight solution. You could still end up with a few too many errors and possible multiple truncation messages but that doesn't seem too bad to me. If you wanted to be completely correct then you could use an `AtomicBoolean`.

```
  private final AtomicBoolean reported = new AtomicBoolean(false);

...
    if (errors.size() < maxErrorsReported) {
      errors.add(String.format(lineErrorMsgFormat, lineErrorMsgArgs));
    } else if (reported.compareAndSet(false, true)) {
      errors.add(
          "Error details truncated. [MaxBadLoadFileLineErrorsReported = "
              + maxErrorsReported
              + "]");
      throw getFormattedException();
    }
```

Reply:
I wanted to leave since the same method can be used in non-concurrent environment

Reply:
> I wanted to leave since the same method can be used in non-concurrent environment

Not sure what you mean by this? The code I suggested will work in both concurrent and non-concurrent environments. Using `synchronized` at the top of the method (which has the same effect as making the method itself `synchronized`, which would be a clearer way to write this if you still want to synchronize in this way) will cause every call to this method to require synchronization. This is wasteful as the code only needs synchronization in the `else` block.

Regarding `synchronized` in general, in my experience many cases can be converted to use `Atomic` classes or other objects that always support synchronization, which I think are easier to maintain and reason about than writing your own synchronization blocks.

Reply:
I think atomic classes work for the container class. In this case, I think the issue was coming from the fact that you can still mutate the child objects.  In any case, I'll try to look a little more.  My main concern was a near term unblock.  I'd like to investigate again at some point using a pet service account to do this (and avoid the need to parallelism at all)